Select objects by distance from attach point - fixes #1618

bengolds opened the pull request.

GitCop commented:
There were the following issues with your Pull Request

 * Commit: c6ab4f04eb710c42ff8af37ea5d7a8ab90b5f95e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

bengolds commented:
Ach, I need to modify the title and I also was modifying an older version of the script. I'll merge onto the newer version soon.

thestonefox commented:
You're going to need to rebase in the latest master and fix any merge conflicts.

But it may be easier for you to make a new branch from my master and migrate your changes over, just so you don't bring in any old thinigs that should have been commited out.

bengolds commented:
I... think I fixed it. I modified the most recent version of VRTK_InteractTouch.cs, rebased my previous commit up to date, and then brought my new file in and amended the commit.

I'm not sure the network is showing everything properly, or if the second commit that's now in the PR is a problem. The code works, I'm just real unsure about this whole git thing.

thestonefox commented:
looks like there's another merge conflict now :(

Also, you probably need to rebase in the latest master into your branch as your commits went a bit haywire

jWalker92 commented:
`// Interact Touch|Interactions|30040
namespace VRTK
{
    using UnityEngine;
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// Event Payload
    /// </summary>
    /// <param name="controllerIndex">The index of the controller doing the interaction.</param>
    /// <param name="target">The GameObject of the interactable object that is being interacted with by the controller.</param>
    public struct ObjectInteractEventArgs
    {
        public uint controllerIndex;
        public GameObject target;
    }

    /// <summary>
    /// Event Payload
    /// </summary>
    /// <param name="sender">this object</param>
    /// <param name="e"><see cref="ObjectInteractEventArgs"/></param>
    public delegate void ObjectInteractEventHandler(object sender, ObjectInteractEventArgs e);

    /// <summary>
    /// The Interact Touch script is usually applied to a Controller and provides a collider to know when the controller is touching something.
    /// </summary>
    /// <remarks>
    /// Colliders are created for the controller and by default the selected controller SDK will have a set of colliders for the given default controller of that SDK.
    ///
    /// A custom collider can be provided by the Custom Rigidbody Object parameter.
    /// </remarks>
    /// <example>
    /// `VRTK/Examples/005_Controller/BasicObjectGrabbing` demonstrates the highlighting of objects that have the `VRTK_InteractableObject` script added to them to show the ability to highlight interactable objects when they are touched by the controllers.
    /// </example>
    [AddComponentMenu("VRTK/Scripts/Interactions/VRTK_InteractTouch")]
    public class VRTK_InteractTouch : MonoBehaviour
    {
        [Tooltip("An optional GameObject that contains the compound colliders to represent the touching object. If this is empty then the collider will be auto generated at runtime to match the SDK default controller.")]
        public GameObject customColliderContainer;

        /// <summary>
        /// Emitted when a valid object is touched.
        /// </summary>
        public event ObjectInteractEventHandler ControllerTouchInteractableObject;
        /// <summary>
        /// Emitted when a valid object is no longer being touched.
        /// </summary>
        public event ObjectInteractEventHandler ControllerUntouchInteractableObject;

        protected GameObject touchedObject = null;
        protected List<Collider> touchedObjectColliders = new List<Collider>();
        protected List<Collider> touchedObjectActiveColliders = new List<Collider>();
        protected GameObject controllerCollisionDetector;
        protected bool triggerRumble;
        protected bool destroyColliderOnDisable;
        protected bool triggerIsColliding = false;
        protected bool triggerWasColliding = false;
        protected bool rigidBodyForcedActive = false;
        protected Rigidbody touchRigidBody;
        protected Object defaultColliderPrefab;

        public enum MultiTouchSelectionStyle
        {
            MostRecent,
            ClosestOrigin
        }

        public MultiTouchSelectionStyle multiTouchSelectionStyle;

        private HashSet<GameObject> touchedObjects = new HashSet<GameObject>();

        public virtual void OnControllerTouchInteractableObject(ObjectInteractEventArgs e)
        {
            if (ControllerTouchInteractableObject != null)
            {
                ControllerTouchInteractableObject(this, e);
            }
        }

        public virtual void OnControllerUntouchInteractableObject(ObjectInteractEventArgs e)
        {
            if (ControllerUntouchInteractableObject != null)
            {
                ControllerUntouchInteractableObject(this, e);
            }
        }

        public virtual ObjectInteractEventArgs SetControllerInteractEvent(GameObject target)
        {
            ObjectInteractEventArgs e;
            e.controllerIndex = VRTK_DeviceFinder.GetControllerIndex(gameObject);
            e.target = target;
            return e;
        }

        /// <summary>
        /// The ForceTouch method will attempt to force the controller to touch the given game object. This is useful if an object that isn't being touched is required to be grabbed or used as the controller doesn't physically have to be touching it to be forced to interact with it.
        /// </summary>
        /// <param name="obj">The game object to attempt to force touch.</param>
        public virtual void ForceTouch(GameObject obj)
        {
            var objCollider = obj.GetComponentInChildren<Collider>();
            if (objCollider != null)
            {
                OnTriggerStay(objCollider);
            }
        }

        /// <summary>
        /// The GetTouchedObject method returns the current object being touched by the controller.
        /// </summary>
        /// <returns>The game object of what is currently being touched by this controller.</returns>
        public virtual GameObject GetTouchedObject()
        {
            return touchedObject;
        }

        /// <summary>
        /// The IsObjectInteractable method is used to check if a given game object is of type `VRTK_InteractableObject` and whether the object is enabled.
        /// </summary>
        /// <param name="obj">The game object to check to see if it's interactable.</param>
        /// <returns>Is true if the given object is of type `VRTK_InteractableObject`.</returns>
        public virtual bool IsObjectInteractable(GameObject obj)
        {
            if (obj != null)
            {
                var io = obj.GetComponentInParent<VRTK_InteractableObject>();
                if (io != null)
                {
                    if (io.disableWhenIdle && !io.enabled)
                    {
                        return true;
                    }
                    return io.enabled;
                }
            }
            return false;
        }

        /// <summary>
        /// The ToggleControllerRigidBody method toggles the controller's rigidbody's ability to detect collisions. If it is true then the controller rigidbody will collide with other collidable game objects.
        /// </summary>
        /// <param name="state">The state of whether the rigidbody is on or off. `true` toggles the rigidbody on and `false` turns it off.</param>
        /// <param name="forceToggle">Determines if the rigidbody has been forced into it's new state by another script. This can be used to override other non-force settings. Defaults to `false`</param>
        public virtual void ToggleControllerRigidBody(bool state, bool forceToggle = false)
        {
            if (controllerCollisionDetector != null && touchRigidBody != null)
            {
                touchRigidBody.isKinematic = !state;
                rigidBodyForcedActive = forceToggle;
                foreach (var collider in controllerCollisionDetector.GetComponentsInChildren<Collider>())
                {
                    collider.isTrigger = !state;
                }
            }
        }

        /// <summary>
        /// The IsRigidBodyActive method checks to see if the rigidbody on the controller object is active and can affect other rigidbodies in the scene.
        /// </summary>
        /// <returns>Is true if the rigidbody on the controller is currently active and able to affect other scene rigidbodies.</returns>
        public virtual bool IsRigidBodyActive()
        {
            return !touchRigidBody.isKinematic;
        }

        /// <summary>
        /// The IsRigidBodyForcedActive method checks to see if the rigidbody on the controller object has been forced into the active state.
        /// </summary>
        /// <returns>Is true if the rigidbody is active and has been forced into the active state.</returns>
        public virtual bool IsRigidBodyForcedActive()
        {
            return (IsRigidBodyActive() && rigidBodyForcedActive);
        }

        /// <summary>
        /// The ForceStopTouching method will stop the controller from touching an object even if the controller is physically touching the object still.
        /// </summary>
        public virtual void ForceStopTouching()
        {
            if (touchedObject != null)
            {
                StopTouching(touchedObject);
            }
        }

        /// <summary>
        /// The ControllerColliders method retrieves all of the associated colliders on the controller.
        /// </summary>
        /// <returns>An array of colliders that are associated with the controller.</returns>
        public virtual Collider[] ControllerColliders()
        {
            return (controllerCollisionDetector.GetComponents<Collider>().Length > 0 ? controllerCollisionDetector.GetComponents<Collider>() : controllerCollisionDetector.GetComponentsInChildren<Collider>());
        }

        protected virtual void Awake()
        {
            destroyColliderOnDisable = false;
            SDK_BaseController.ControllerHand controllerHand = VRTK_DeviceFinder.GetControllerHand(gameObject);
            defaultColliderPrefab = Resources.Load(VRTK_SDK_Bridge.GetControllerDefaultColliderPath(controllerHand));
        }

        protected virtual void OnEnable()
        {
            VRTK_PlayerObject.SetPlayerObject(gameObject, VRTK_PlayerObject.ObjectTypes.Controller);
            triggerRumble = false;
            CreateTouchCollider();
            CreateTouchRigidBody();
        }

        protected virtual void OnDisable()
        {
            ForceStopTouching();
            DestroyTouchCollider();
        }

        protected virtual void OnTriggerEnter(Collider collider)
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                GameObject colliderInteractableObject = TriggerStart(collider);
                //If the new collider is not part of the existing touched object (and the object isn't being grabbed) then start touching the new object
                if (touchedObject != null && colliderInteractableObject && touchedObject != colliderInteractableObject && !touchedObject.GetComponent<VRTK_InteractableObject>().IsGrabbed())
                {
                    CancelInvoke("ResetTriggerRumble");
                    ResetTriggerRumble();
                    ForceStopTouching();
                    triggerIsColliding = true;
                }
            }
            else if (multiTouchSelectionStyle == MultiTouchSelectionStyle.ClosestOrigin)
            {
                if (!collider.gameObject.name.StartsWith("Hull"))
                {
                    if (IsObjectInteractable(collider.gameObject))
                    {
                        touchedObjects.Add(collider.gameObject);
                    }
                }
                //Debug.Log("Added " + collider.name + " " + touchedObjects.Count);
            }
        }

        protected virtual void OnTriggerExit(Collider collider)
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                if (touchedObjectActiveColliders.Contains(collider))
                {
                    touchedObjectActiveColliders.Remove(collider);
                }
            }
            else if (multiTouchSelectionStyle == MultiTouchSelectionStyle.ClosestOrigin)
            {
                if (!collider.gameObject.name.StartsWith("Hull"))
                {
                    if (touchedObject == collider.gameObject)
                    {
                        StopTouching(collider.gameObject);
                    }
                    touchedObjects.Remove(collider.gameObject);
                }
                //Debug.Log("Removed " + collider.name + " " + touchedObjects.Count);
            }
        }

        protected virtual void OnTriggerStay(Collider collider)
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                GameObject colliderInteractableObject = TriggerStart(collider);

                if (touchedObject == null || collider.transform.IsChildOf(touchedObject.transform))
                {
                    triggerIsColliding = true;
                }

                if (touchedObject == null && colliderInteractableObject && IsObjectInteractable(collider.gameObject))
                {
                    touchedObject = colliderInteractableObject;
                    var touchedObjectScript = touchedObject.GetComponent<VRTK_InteractableObject>();
                    GameObject touchingObject = gameObject;

                    //If this controller is not allowed to touch this interactable object then clean up touch and return before initiating a touch.
                    if (!touchedObjectScript.IsValidInteractableController(gameObject, touchedObjectScript.allowedTouchControllers))
                    {
                        CleanupEndTouch();
                        return;
                    }
                    StoreTouchedObjectColliders(collider);

                    touchedObjectScript.ToggleHighlight(true);
                    ToggleControllerVisibility(false);
                    CheckRumbleController(touchedObjectScript);
                    touchedObjectScript.StartTouching(touchingObject);

                    OnControllerTouchInteractableObject(SetControllerInteractEvent(touchedObject));
                }
            }
        }

        protected virtual void FixedUpdate()
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                if (!triggerIsColliding && !triggerWasColliding)
                {
                    CheckStopTouching();
                }
                triggerWasColliding = triggerIsColliding;
                triggerIsColliding = false;
            }
        }

        protected virtual void LateUpdate()
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.ClosestOrigin)
            {
                touchedObjects.RemoveWhere(element => element == null);
                if (touchedObjects.Count > 0)
                {
                    //TODO: Maybe make this clearer?
                    var grabScript = GetComponent<VRTK_InteractGrab>();
                    var controllerOrigin = grabScript != null && grabScript.controllerAttachPoint != null ?
                                            grabScript.controllerAttachPoint.transform.position :
                                            transform.position;
                    var closestTouched = touchedObjects.OrderBy(element => Vector3.Distance(controllerOrigin, element.transform.position)).FirstOrDefault();
                    if (touchedObject != closestTouched)
                    {
                        if (touchedObject != null)
                        {
                            StopTouching(touchedObject);
                        }
                        StartTouching(closestTouched);
                    }

                }
            }
            else if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                if (touchedObjectActiveColliders.Count == 0)
                {
                    CheckStopTouching();
                }
            }
        }

        private void StartTouching(GameObject obj)
        {
            if (obj.GetComponent<VRTK_InteractableObject>())
            {
                touchedObject = obj;
            }
            else
            {
                touchedObject = obj.GetComponentInParent<VRTK_InteractableObject>().gameObject;
            }

            var touchedObjectScript = touchedObject.GetComponent<VRTK_InteractableObject>();

            if (!touchedObjectScript.IsValidInteractableController(gameObject, touchedObjectScript.allowedTouchControllers))
            {
                touchedObject = null;
                return;
            }

            touchedObjectScript.ToggleHighlight(true);
            ToggleControllerVisibility(false);
            CheckRumbleController(touchedObjectScript);
            touchedObjectScript.StartTouching(gameObject);

            OnControllerTouchInteractableObject(SetControllerInteractEvent(touchedObject));
        }

        protected virtual GameObject GetColliderInteractableObject(Collider collider)
        {
            var checkIO = collider.GetComponentInParent<VRTK_InteractableObject>();
            return (checkIO ? checkIO.gameObject : null);
        }

        protected virtual void AddActiveCollider(Collider collider)
        {
            if (touchedObject != null && !touchedObjectActiveColliders.Contains(collider) && touchedObjectColliders.Contains(collider))
            {
                touchedObjectActiveColliders.Add(collider);
            }
        }

        protected virtual void StoreTouchedObjectColliders(Collider collider)
        {
            touchedObjectColliders.Clear();
            touchedObjectActiveColliders.Clear();
            foreach (var touchedObjectCollider in touchedObject.GetComponentsInChildren<Collider>())
            {
                touchedObjectColliders.Add(touchedObjectCollider);
            }
            touchedObjectActiveColliders.Add(collider);
        }

        protected virtual void ToggleControllerVisibility(bool visible)
        {
            GameObject modelContainer = VRTK_DeviceFinder.GetModelAliasController(gameObject);
            if (touchedObject != null)
            {
                VRTK_InteractControllerAppearance[] controllerAppearanceScript = touchedObject.GetComponentsInParent<VRTK_InteractControllerAppearance>(true);
                if (controllerAppearanceScript.Length > 0)
                {
                    controllerAppearanceScript[0].ToggleControllerOnTouch(visible, modelContainer, touchedObject);
                }
            }
            else if (visible)
            {
                VRTK_SharedMethods.SetRendererVisible(modelContainer, touchedObject);
            }
        }

        protected virtual void CheckRumbleController(VRTK_InteractableObject touchedObjectScript)
        {
            if (!triggerRumble)
            {
                var doHaptics = touchedObject.GetComponentInParent<VRTK_InteractHaptics>();
                if (doHaptics != null)
                {
                    triggerRumble = true;
                    doHaptics.HapticsOnTouch(VRTK_DeviceFinder.GetControllerIndex(gameObject));
                    Invoke("ResetTriggerRumble", doHaptics.durationOnTouch);
                }
            }
        }

        protected virtual void CheckStopTouching()
        {
            if (touchedObject != null)
            {
                var touchedObjectScript = touchedObject.GetComponent<VRTK_InteractableObject>();
                GameObject touchingObject = gameObject;

                //If it's being grabbed by the current touching object then it hasn't stopped being touched.
                if (touchedObjectScript != null && touchedObjectScript.GetGrabbingObject() != touchingObject)
                {
                    StopTouching(touchedObject);
                }
            }
        }

        protected virtual GameObject TriggerStart(Collider collider)
        {
            if (IsSnapDropZone(collider))
            {
                return null;
            }

            AddActiveCollider(collider);

            return GetColliderInteractableObject(collider);
        }

        protected virtual bool IsSnapDropZone(Collider collider)
        {
            if (collider.GetComponent<VRTK_SnapDropZone>())
            {
                return true;
            }
            return false;
        }

        protected virtual void ResetTriggerRumble()
        {
            triggerRumble = false;
        }

        protected virtual void StopTouching(GameObject untouched)
        {
            if (IsObjectInteractable(untouched))
            {
                GameObject touchingObject = gameObject;
                var untouchedObjectScript = untouched.GetComponent<VRTK_InteractableObject>();
                untouchedObjectScript.StopTouching(touchingObject);
                if (!untouchedObjectScript.IsTouched())
                {
                    untouchedObjectScript.ToggleHighlight(false);
                }
            }

            ToggleControllerVisibility(true);
            OnControllerUntouchInteractableObject(SetControllerInteractEvent(untouched));
            CleanupEndTouch();
        }

        protected virtual void CleanupEndTouch()
        {
            touchedObject = null;
            touchedObjectActiveColliders.Clear();
            touchedObjectColliders.Clear();
        }

        protected virtual void DestroyTouchCollider()
        {
            if (destroyColliderOnDisable)
            {
                Destroy(controllerCollisionDetector);
            }
        }

        protected virtual bool CustomRigidBodyIsChild()
        {
            foreach (var childTransform in GetComponentsInChildren<Transform>())
            {
                if (childTransform != transform && childTransform == customColliderContainer.transform)
                {
                    return true;
                }
            }
            return false;
        }

        protected virtual void CreateTouchCollider()
        {
            if (customColliderContainer == null)
            {
                if (!defaultColliderPrefab)
                {
                    VRTK_Logger.Error(VRTK_Logger.GetCommonMessage(VRTK_Logger.CommonMessageKeys.SDK_OBJECT_NOT_FOUND, "default collider prefab", "Controller SDK"));
                    return;
                }
                controllerCollisionDetector = Instantiate(defaultColliderPrefab, transform.position, transform.rotation) as GameObject;
                controllerCollisionDetector.transform.SetParent(transform);
                controllerCollisionDetector.transform.localScale = transform.localScale;
                controllerCollisionDetector.name = VRTK_SharedMethods.GenerateVRTKObjectName(true, "Controller", "CollidersContainer");
                destroyColliderOnDisable = true;
            }
            else
            {
                if (CustomRigidBodyIsChild())
                {
                    controllerCollisionDetector = customColliderContainer;
                    destroyColliderOnDisable = false;
                }
                else
                {
                    controllerCollisionDetector = Instantiate(customColliderContainer, transform.position, transform.rotation) as GameObject;
                    controllerCollisionDetector.transform.SetParent(transform);
                    controllerCollisionDetector.transform.localScale = transform.localScale;
                    destroyColliderOnDisable = true;
                }
            }
            controllerCollisionDetector.AddComponent<VRTK_PlayerObject>().objectType = VRTK_PlayerObject.ObjectTypes.Collider;
        }

        protected virtual void CreateTouchRigidBody()
        {
            touchRigidBody = GetComponent<Rigidbody>();
            if (touchRigidBody == null)
            {
                touchRigidBody = gameObject.AddComponent<Rigidbody>();
                touchRigidBody.isKinematic = true;
                touchRigidBody.useGravity = false;
                touchRigidBody.constraints = RigidbodyConstraints.FreezeAll;
                touchRigidBody.collisionDetectionMode = CollisionDetectionMode.ContinuousDynamic;
            }
        }
    }
}`

_Seems_ to work

jWalker92 commented:
`
// Interact Touch|Interactions|30040
namespace VRTK
{
    using UnityEngine;
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// Event Payload
    /// </summary>
    /// <param name="controllerIndex">The index of the controller doing the interaction.</param>
    /// <param name="target">The GameObject of the interactable object that is being interacted with by the controller.</param>
    public struct ObjectInteractEventArgs
    {
        public uint controllerIndex;
        public GameObject target;
    }

    /// <summary>
    /// Event Payload
    /// </summary>
    /// <param name="sender">this object</param>
    /// <param name="e"><see cref="ObjectInteractEventArgs"/></param>
    public delegate void ObjectInteractEventHandler(object sender, ObjectInteractEventArgs e);

    /// <summary>
    /// The Interact Touch script is usually applied to a Controller and provides a collider to know when the controller is touching something.
    /// </summary>
    /// <remarks>
    /// Colliders are created for the controller and by default the selected controller SDK will have a set of colliders for the given default controller of that SDK.
    ///
    /// A custom collider can be provided by the Custom Rigidbody Object parameter.
    /// </remarks>
    /// <example>
    /// `VRTK/Examples/005_Controller/BasicObjectGrabbing` demonstrates the highlighting of objects that have the `VRTK_InteractableObject` script added to them to show the ability to highlight interactable objects when they are touched by the controllers.
    /// </example>
    [AddComponentMenu("VRTK/Scripts/Interactions/VRTK_InteractTouch")]
    public class VRTK_InteractTouch : MonoBehaviour
    {
        [Tooltip("An optional GameObject that contains the compound colliders to represent the touching object. If this is empty then the collider will be auto generated at runtime to match the SDK default controller.")]
        public GameObject customColliderContainer;

        /// <summary>
        /// Emitted when a valid object is touched.
        /// </summary>
        public event ObjectInteractEventHandler ControllerTouchInteractableObject;
        /// <summary>
        /// Emitted when a valid object is no longer being touched.
        /// </summary>
        public event ObjectInteractEventHandler ControllerUntouchInteractableObject;

        protected GameObject touchedObject = null;
        protected List<Collider> touchedObjectColliders = new List<Collider>();
        protected List<Collider> touchedObjectActiveColliders = new List<Collider>();
        protected GameObject controllerCollisionDetector;
        protected bool triggerRumble;
        protected bool destroyColliderOnDisable;
        protected bool triggerIsColliding = false;
        protected bool triggerWasColliding = false;
        protected bool rigidBodyForcedActive = false;
        protected Rigidbody touchRigidBody;
        protected Object defaultColliderPrefab;

        public enum MultiTouchSelectionStyle
        {
            MostRecent,
            ClosestOrigin
        }

        public MultiTouchSelectionStyle multiTouchSelectionStyle;

        private HashSet<GameObject> touchedObjects = new HashSet<GameObject>();

        public virtual void OnControllerTouchInteractableObject(ObjectInteractEventArgs e)
        {
            if (ControllerTouchInteractableObject != null)
            {
                ControllerTouchInteractableObject(this, e);
            }
        }

        public virtual void OnControllerUntouchInteractableObject(ObjectInteractEventArgs e)
        {
            if (ControllerUntouchInteractableObject != null)
            {
                ControllerUntouchInteractableObject(this, e);
            }
        }

        public virtual ObjectInteractEventArgs SetControllerInteractEvent(GameObject target)
        {
            ObjectInteractEventArgs e;
            e.controllerIndex = VRTK_DeviceFinder.GetControllerIndex(gameObject);
            e.target = target;
            return e;
        }

        /// <summary>
        /// The ForceTouch method will attempt to force the controller to touch the given game object. This is useful if an object that isn't being touched is required to be grabbed or used as the controller doesn't physically have to be touching it to be forced to interact with it.
        /// </summary>
        /// <param name="obj">The game object to attempt to force touch.</param>
        public virtual void ForceTouch(GameObject obj)
        {
            var objCollider = obj.GetComponentInChildren<Collider>();
            if (objCollider != null)
            {
                OnTriggerStay(objCollider);
            }
        }

        /// <summary>
        /// The GetTouchedObject method returns the current object being touched by the controller.
        /// </summary>
        /// <returns>The game object of what is currently being touched by this controller.</returns>
        public virtual GameObject GetTouchedObject()
        {
            return touchedObject;
        }

        /// <summary>
        /// The IsObjectInteractable method is used to check if a given game object is of type `VRTK_InteractableObject` and whether the object is enabled.
        /// </summary>
        /// <param name="obj">The game object to check to see if it's interactable.</param>
        /// <returns>Is true if the given object is of type `VRTK_InteractableObject`.</returns>
        public virtual bool IsObjectInteractable(GameObject obj)
        {
            if (obj != null)
            {
                var io = obj.GetComponentInParent<VRTK_InteractableObject>();
                if (io != null)
                {
                    if (io.disableWhenIdle && !io.enabled)
                    {
                        return true;
                    }
                    return io.enabled;
                }
            }
            return false;
        }

        /// <summary>
        /// The ToggleControllerRigidBody method toggles the controller's rigidbody's ability to detect collisions. If it is true then the controller rigidbody will collide with other collidable game objects.
        /// </summary>
        /// <param name="state">The state of whether the rigidbody is on or off. `true` toggles the rigidbody on and `false` turns it off.</param>
        /// <param name="forceToggle">Determines if the rigidbody has been forced into it's new state by another script. This can be used to override other non-force settings. Defaults to `false`</param>
        public virtual void ToggleControllerRigidBody(bool state, bool forceToggle = false)
        {
            if (controllerCollisionDetector != null && touchRigidBody != null)
            {
                touchRigidBody.isKinematic = !state;
                rigidBodyForcedActive = forceToggle;
                foreach (var collider in controllerCollisionDetector.GetComponentsInChildren<Collider>())
                {
                    collider.isTrigger = !state;
                }
            }
        }

        /// <summary>
        /// The IsRigidBodyActive method checks to see if the rigidbody on the controller object is active and can affect other rigidbodies in the scene.
        /// </summary>
        /// <returns>Is true if the rigidbody on the controller is currently active and able to affect other scene rigidbodies.</returns>
        public virtual bool IsRigidBodyActive()
        {
            return !touchRigidBody.isKinematic;
        }

        /// <summary>
        /// The IsRigidBodyForcedActive method checks to see if the rigidbody on the controller object has been forced into the active state.
        /// </summary>
        /// <returns>Is true if the rigidbody is active and has been forced into the active state.</returns>
        public virtual bool IsRigidBodyForcedActive()
        {
            return (IsRigidBodyActive() && rigidBodyForcedActive);
        }

        /// <summary>
        /// The ForceStopTouching method will stop the controller from touching an object even if the controller is physically touching the object still.
        /// </summary>
        public virtual void ForceStopTouching()
        {
            if (touchedObject != null)
            {
                StopTouching(touchedObject);
            }
        }

        /// <summary>
        /// The ControllerColliders method retrieves all of the associated colliders on the controller.
        /// </summary>
        /// <returns>An array of colliders that are associated with the controller.</returns>
        public virtual Collider[] ControllerColliders()
        {
            return (controllerCollisionDetector.GetComponents<Collider>().Length > 0 ? controllerCollisionDetector.GetComponents<Collider>() : controllerCollisionDetector.GetComponentsInChildren<Collider>());
        }

        protected virtual void Awake()
        {
            destroyColliderOnDisable = false;
            SDK_BaseController.ControllerHand controllerHand = VRTK_DeviceFinder.GetControllerHand(gameObject);
            defaultColliderPrefab = Resources.Load(VRTK_SDK_Bridge.GetControllerDefaultColliderPath(controllerHand));
        }

        protected virtual void OnEnable()
        {
            VRTK_PlayerObject.SetPlayerObject(gameObject, VRTK_PlayerObject.ObjectTypes.Controller);
            triggerRumble = false;
            CreateTouchCollider();
            CreateTouchRigidBody();
        }

        protected virtual void OnDisable()
        {
            ForceStopTouching();
            DestroyTouchCollider();
        }

        protected virtual void OnTriggerEnter(Collider collider)
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                GameObject colliderInteractableObject = TriggerStart(collider);
                //If the new collider is not part of the existing touched object (and the object isn't being grabbed) then start touching the new object
                if (touchedObject != null && colliderInteractableObject && touchedObject != colliderInteractableObject && !touchedObject.GetComponent<VRTK_InteractableObject>().IsGrabbed())
                {
                    CancelInvoke("ResetTriggerRumble");
                    ResetTriggerRumble();
                    ForceStopTouching();
                    triggerIsColliding = true;
                }
            }
            else if (multiTouchSelectionStyle == MultiTouchSelectionStyle.ClosestOrigin)
            {
                    if (IsObjectInteractable(collider.gameObject))
                    {
                        touchedObjects.Add(collider.gameObject);
                    }
                //Debug.Log("Added " + collider.name + " " + touchedObjects.Count);
            }
        }

        protected virtual void OnTriggerExit(Collider collider)
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                if (touchedObjectActiveColliders.Contains(collider))
                {
                    touchedObjectActiveColliders.Remove(collider);
                }
            }
            else if (multiTouchSelectionStyle == MultiTouchSelectionStyle.ClosestOrigin)
            {
                    if (touchedObject == collider.gameObject)
                    {
                        StopTouching(collider.gameObject);
                    }
                    touchedObjects.Remove(collider.gameObject);
                //Debug.Log("Removed " + collider.name + " " + touchedObjects.Count);
            }
        }

        protected virtual void OnTriggerStay(Collider collider)
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                GameObject colliderInteractableObject = TriggerStart(collider);

                if (touchedObject == null || collider.transform.IsChildOf(touchedObject.transform))
                {
                    triggerIsColliding = true;
                }

                if (touchedObject == null && colliderInteractableObject && IsObjectInteractable(collider.gameObject))
                {
                    touchedObject = colliderInteractableObject;
                    var touchedObjectScript = touchedObject.GetComponent<VRTK_InteractableObject>();
                    GameObject touchingObject = gameObject;

                    //If this controller is not allowed to touch this interactable object then clean up touch and return before initiating a touch.
                    if (!touchedObjectScript.IsValidInteractableController(gameObject, touchedObjectScript.allowedTouchControllers))
                    {
                        CleanupEndTouch();
                        return;
                    }
                    StoreTouchedObjectColliders(collider);

                    touchedObjectScript.ToggleHighlight(true);
                    ToggleControllerVisibility(false);
                    CheckRumbleController(touchedObjectScript);
                    touchedObjectScript.StartTouching(touchingObject);

                    OnControllerTouchInteractableObject(SetControllerInteractEvent(touchedObject));
                }
            }
        }

        protected virtual void FixedUpdate()
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                if (!triggerIsColliding && !triggerWasColliding)
                {
                    CheckStopTouching();
                }
                triggerWasColliding = triggerIsColliding;
                triggerIsColliding = false;
            }
        }

        protected virtual void LateUpdate()
        {
            if (multiTouchSelectionStyle == MultiTouchSelectionStyle.ClosestOrigin)
            {
                touchedObjects.RemoveWhere(element => element == null);
                if (touchedObjects.Count > 0)
                {
                    //TODO: Maybe make this clearer?
                    var grabScript = GetComponent<VRTK_InteractGrab>();
                    var controllerOrigin = grabScript != null && grabScript.controllerAttachPoint != null ?
                                            grabScript.controllerAttachPoint.transform.position :
                                            transform.position;
                    var closestTouched = touchedObjects.OrderBy(element => Vector3.Distance(controllerOrigin, element.transform.position)).FirstOrDefault();
                    if (touchedObject != closestTouched)
                    {
                        if (touchedObject != null)
                        {
                            StopTouching(touchedObject);
                        }
                        StartTouching(closestTouched);
                    }

                }
            }
            else if (multiTouchSelectionStyle == MultiTouchSelectionStyle.MostRecent)
            {
                if (touchedObjectActiveColliders.Count == 0)
                {
                    CheckStopTouching();
                }
            }
        }

        private void StartTouching(GameObject obj)
        {
            if (obj.GetComponent<VRTK_InteractableObject>())
            {
                touchedObject = obj;
            }
            else
            {
                touchedObject = obj.GetComponentInParent<VRTK_InteractableObject>().gameObject;
            }

            var touchedObjectScript = touchedObject.GetComponent<VRTK_InteractableObject>();

            if (!touchedObjectScript.IsValidInteractableController(gameObject, touchedObjectScript.allowedTouchControllers))
            {
                touchedObject = null;
                return;
            }

            touchedObjectScript.ToggleHighlight(true);
            ToggleControllerVisibility(false);
            CheckRumbleController(touchedObjectScript);
            touchedObjectScript.StartTouching(gameObject);

            OnControllerTouchInteractableObject(SetControllerInteractEvent(touchedObject));
        }

        protected virtual GameObject GetColliderInteractableObject(Collider collider)
        {
            var checkIO = collider.GetComponentInParent<VRTK_InteractableObject>();
            return (checkIO ? checkIO.gameObject : null);
        }

        protected virtual void AddActiveCollider(Collider collider)
        {
            if (touchedObject != null && !touchedObjectActiveColliders.Contains(collider) && touchedObjectColliders.Contains(collider))
            {
                touchedObjectActiveColliders.Add(collider);
            }
        }

        protected virtual void StoreTouchedObjectColliders(Collider collider)
        {
            touchedObjectColliders.Clear();
            touchedObjectActiveColliders.Clear();
            foreach (var touchedObjectCollider in touchedObject.GetComponentsInChildren<Collider>())
            {
                touchedObjectColliders.Add(touchedObjectCollider);
            }
            touchedObjectActiveColliders.Add(collider);
        }

        protected virtual void ToggleControllerVisibility(bool visible)
        {
            GameObject modelContainer = VRTK_DeviceFinder.GetModelAliasController(gameObject);
            if (touchedObject != null)
            {
                VRTK_InteractControllerAppearance[] controllerAppearanceScript = touchedObject.GetComponentsInParent<VRTK_InteractControllerAppearance>(true);
                if (controllerAppearanceScript.Length > 0)
                {
                    controllerAppearanceScript[0].ToggleControllerOnTouch(visible, modelContainer, touchedObject);
                }
            }
            else if (visible)
            {
                VRTK_SharedMethods.SetRendererVisible(modelContainer, touchedObject);
            }
        }

        protected virtual void CheckRumbleController(VRTK_InteractableObject touchedObjectScript)
        {
            if (!triggerRumble)
            {
                var doHaptics = touchedObject.GetComponentInParent<VRTK_InteractHaptics>();
                if (doHaptics != null)
                {
                    triggerRumble = true;
                    doHaptics.HapticsOnTouch(VRTK_DeviceFinder.GetControllerIndex(gameObject));
                    Invoke("ResetTriggerRumble", doHaptics.durationOnTouch);
                }
            }
        }

        protected virtual void CheckStopTouching()
        {
            if (touchedObject != null)
            {
                var touchedObjectScript = touchedObject.GetComponent<VRTK_InteractableObject>();
                GameObject touchingObject = gameObject;

                //If it's being grabbed by the current touching object then it hasn't stopped being touched.
                if (touchedObjectScript != null && touchedObjectScript.GetGrabbingObject() != touchingObject)
                {
                    StopTouching(touchedObject);
                }
            }
        }

        protected virtual GameObject TriggerStart(Collider collider)
        {
            if (IsSnapDropZone(collider))
            {
                return null;
            }

            AddActiveCollider(collider);

            return GetColliderInteractableObject(collider);
        }

        protected virtual bool IsSnapDropZone(Collider collider)
        {
            if (collider.GetComponent<VRTK_SnapDropZone>())
            {
                return true;
            }
            return false;
        }

        protected virtual void ResetTriggerRumble()
        {
            triggerRumble = false;
        }

        protected virtual void StopTouching(GameObject untouched)
        {
            if (IsObjectInteractable(untouched))
            {
                GameObject touchingObject = gameObject;
                var untouchedObjectScript = untouched.GetComponent<VRTK_InteractableObject>();
                untouchedObjectScript.StopTouching(touchingObject);
                if (!untouchedObjectScript.IsTouched())
                {
                    untouchedObjectScript.ToggleHighlight(false);
                }
            }

            ToggleControllerVisibility(true);
            OnControllerUntouchInteractableObject(SetControllerInteractEvent(untouched));
            CleanupEndTouch();
        }

        protected virtual void CleanupEndTouch()
        {
            touchedObject = null;
            touchedObjectActiveColliders.Clear();
            touchedObjectColliders.Clear();
        }

        protected virtual void DestroyTouchCollider()
        {
            if (destroyColliderOnDisable)
            {
                Destroy(controllerCollisionDetector);
            }
        }

        protected virtual bool CustomRigidBodyIsChild()
        {
            foreach (var childTransform in GetComponentsInChildren<Transform>())
            {
                if (childTransform != transform && childTransform == customColliderContainer.transform)
                {
                    return true;
                }
            }
            return false;
        }

        protected virtual void CreateTouchCollider()
        {
            if (customColliderContainer == null)
            {
                if (!defaultColliderPrefab)
                {
                    VRTK_Logger.Error(VRTK_Logger.GetCommonMessage(VRTK_Logger.CommonMessageKeys.SDK_OBJECT_NOT_FOUND, "default collider prefab", "Controller SDK"));
                    return;
                }
                controllerCollisionDetector = Instantiate(defaultColliderPrefab, transform.position, transform.rotation) as GameObject;
                controllerCollisionDetector.transform.SetParent(transform);
                controllerCollisionDetector.transform.localScale = transform.localScale;
                controllerCollisionDetector.name = VRTK_SharedMethods.GenerateVRTKObjectName(true, "Controller", "CollidersContainer");
                destroyColliderOnDisable = true;
            }
            else
            {
                if (CustomRigidBodyIsChild())
                {
                    controllerCollisionDetector = customColliderContainer;
                    destroyColliderOnDisable = false;
                }
                else
                {
                    controllerCollisionDetector = Instantiate(customColliderContainer, transform.position, transform.rotation) as GameObject;
                    controllerCollisionDetector.transform.SetParent(transform);
                    controllerCollisionDetector.transform.localScale = transform.localScale;
                    destroyColliderOnDisable = true;
                }
            }
            controllerCollisionDetector.AddComponent<VRTK_PlayerObject>().objectType = VRTK_PlayerObject.ObjectTypes.Collider;
        }

        protected virtual void CreateTouchRigidBody()
        {
            touchRigidBody = GetComponent<Rigidbody>();
            if (touchRigidBody == null)
            {
                touchRigidBody = gameObject.AddComponent<Rigidbody>();
                touchRigidBody.isKinematic = true;
                touchRigidBody.useGravity = false;
                touchRigidBody.constraints = RigidbodyConstraints.FreezeAll;
                touchRigidBody.collisionDetectionMode = CollisionDetectionMode.ContinuousDynamic;
            }
        }
    }
}
`

I just changed the current version i downloaded from gitHub with the changes from this PR and it _Seems_ to work. It has some problems with many colliders on one object at the time. But I will see if i can use layers for touchableZoneColliders so that the actual radius is more controllable in that you can grab objects.

FreakinKat commented:
Ugh what a mess y'all are confusing me on this, is this an ongoing problem? I see it's been a year and it's not Auto correcting itself, who wrote the script? Rewrite it and tell it to update itself for you, all ways remember if/then or c'mon fellas. But if you'd like I can write it up for you that way when you get these errors it searches itself for the problem, identifys, you can set the possibility of solutions search (add solutions here) if needed have it set to search for pull or search for packet, then load solution and if correct it rewrites with functional line and proceeds with out any delay. Shouldn't be that obfuscated to make a little list of plasable solution lines, or so it doesn't get all stupid from people fucking with it just make a set line that will be the next updated versions set line and have it bump itself set an infinite string which you can basically pull from Google and have it set to I donno power 500,000 which would probably end in the year 2,253,975,060,360,145,273,856,249,228,679,335,484,114,458,988,044 .... honestly do I need to continue because by that year we'll probably have been turned to dust universe exploded and on a 4 or 5th deminsional at which point VR and 3D will be less then 8bit and we'll probably all be God's or something running our own universe and keeping some form of reality stable in time and space because we're bored with other dumb god stuff that well probably be doing, but at least you'll know that in the year 2017 when your physical bodys mind was stuck on a problem you'd have it fixed for many years proceeding itself... I donno I've never been a god before ... Or to my knowledge I don't think so. That or I made sure I'd forget that I was and had the main god remake me to forget that I was a god but to remember to help some dudes out with some simple problem they made confusing when it didn't have to be... I donno I'm basically questioning my own existence at this point so ill either permanently fix this for y'all or someone else do it, I'm busy trying to merge a couple of corporate ceo retards to keep all these fucking licensing agreements from increasing because everyone wants to be the big dick pulling all the strings and that's fucking stupid either they play nice or I'll erase them and their industrial financial records and encrypt their entire client base into a locked zip and smash the hard drive with a hammer and toast it in a smelting pot and make a fucking necklace out of the material that usto be their billion dollar paycheck god damn it!!! So u can see I'm a little frustrated how such a simple problem threw me off. It's not y'all's fault I'm just trying to get into this whole unity thing with the VR stuff only to find out it's basically drag and drop prog level coding I had no idea it was so fucking god damn simple I could have probably taught my 🐈 the fundimentals by now but all these companies want is licensed this and licensed that everyone wants to make shit but then when people want to use it for something and make a little money for themselves to eat and feed their cat and fix their cars ALLLLLLLLL of a fucking sudden everyone wants to drag legal boundaries into everythinf
 Well No, I usto be into super serious black hat shit and I'm really starting to question rather I want to totally fuck up the world markets online that can't get the fucking hell along with each other, if Steve was still here he'd have agreed along time ago and none of this would have been as annoying as it is. And government wonders why people arnt as gullible as it usto be before they realized they fucked them selves with the internet and books, our forfathers wouldn't have been called Anonymous they'd have been called beatyourass squad and over thrown the government by now. But everyones pussy footing around all scared and shit like they're gonna get exploded by some terroist that are just retards paying retards to keep fear and fake protection from imaginary evil terrorist that just magically appeared out of thin air and where super sneaky and bombing this and gunning down that... Cmon seriously 😒 like the feds and Homeland security are just that fucking benine to terrorist activity that they don't know when a fucking grenade launcher thing or fully automatic assualt rifles and 20,000boxes of ammo gos missing or gets imported into our county get fucking serious their supposed to keep track of this shit yet here we are having New York getting jacked up and fake ass medias blantently lying to the extremely few who still believes news media anymore. Their all Done they're old they are literally to fucking old to keep up with how complex the world has become and kids are basically exponentially geniuses and literally have the entire worlds knowledge at their fingertips. Give prisoners computers with the internet and give them 5years you'll have alot of fucking murderers getting out for a formality  or misspelling in there trial. These old ass power hungry idiots are old out dated skeletons and their children and families want them to die so they can get all daddy's money while feminist bull dikes start controlling shit, which they are on the cool and it be nice if they had a cool sweet job as a sex slave or something for their hot daughters who are around my age 29by the way (for you tech lady's out there, don't worry I'm not completely insane plus I know how to appreciate a women *just saying, don't knock this one off until you test drive him* just saying** ) I'm totally ranting and getting alot of stress off my chest right now so I think that's totally awesome and cool and I get to be apart of this community Yippy 😸💪 but don't worry I'm harmless seriously I'm like one of those fuzzy Pusheen cats only not all fat and stubby but I'm no body builder lmao 😂. And for the record I'll go down and say this for y'all VR guys y'all are such a respectable bunch of winners I completely admire y'all so young and already so far into understanding computers and sharing your knowledge freely and openly for all, it's what us older heads who still fight the good fight for freedom love to see, I'm not that old either ya little turds! So shut up 29s the new 21 so shhhhhh. But seriously for all you other engineers and developer masters y'all should take some notes from the VR community and humble yourselves y'all ain't gods of poop. I've got 15+yrs learning codes out the ass and history y'all are dumb for spending all that money learning what you could have learned at home on your computer for free and written your own computer language and been millionaires had you the brain to think like a genius 😹 lol, I got my first Rolex at 16 and my family's middle-class and I was just a broke ass punkrocker with a PC but I still drove a BMW and have a Rolex which I still have ty very much Intel don't worry your secrets safe with me (no seriously it still is don't have me killed) but yah y'all VR guys are great man don't let these ethics"white hat hackers" bully y'all and make you think your not safe and all your hard work is at risk of being stolen all them dicks do is claim to be good and evil black hats are out to get you boogie boogie boo ahh the evil black hat hackers gonna get you... Well I usto be one and I doubt I seem scary or seem like a monster now do I? At most your probably like this guy's crazy , and you'd probably be correct hence the old Nick name crazy matt , but hey whatever I'm just very honest and blunt about how I feel about shit, I mean fuck all those mind games and shit it only makes life difficult and annoying and then people don't trust each other no more. Well  have no fear ok :) hell if anything look up lizard squad and look what I did to those bunch of losers 😹 haha yeah that was me, but they shouldn't have fucked with us gamers much less DDOSing me without giving me a heads up then calling me a bitch when I told that one idiot of theirs to chill. Where y'all lizard squad peeps at now? Clowns, If you ask me, anyways I'll read up on this later thanks whoever wrote this up it be nice when I can join a cool group and we'll make a sick mmo vr 🤤omg it'll be sexy, but just give me a holler if someone tries to threaten y'all with something I'll take care of them free of charge. Seriously don't let some one threaten y'all with encrypting your hard work and making you pay them to get it back or anything like that, I'll get it back for you and make them apologize (yes that's a challenge to all you badasses out there) I'll fuck your world up like Rick from Rick and Morty "I'm like Dr who in this motherfucker!" Oh and Donald Trump's did gay stuff lol. Ha blame the church on that leak lmao🤣. Alrighty thank y'all for letting me post here I feel great, I might just be nice to those ceo dudes that I'll be talking to in an hour 😉 we'll keep VR stress free so keep up the good work guys and gals we love y'all and some of us look forward to joining y'all's groups once we've delt with these business types trying to give y'all a hard time. By the way this was written by 4people claiming the same ISP DNS ip of FreakinKat (sorry bro we love you, we knew you wouldn't mind) 


Your right I don't mind David but it's freaking 5:57in the morning u noob and why are you remotely using my phone dude it's stupid

Your PC's offline
Yeah a fucking capassitor got knocked off from a video card 
Shkrnrjdusbeb no shut up I don't want to djdjdndmd
ND no stop I'm not letting you say it snddndjdndd
Dkbjb hush damn you I'm posting this as is d dkjd d d.   Lol I'll do it bro.f you I'm posting it

RunenGamer commented:
@thestonefox : Can be closed, since it was reopened with a new PR (#1811).

thestonefox commented:
I'm going to start closing these, we can reopen them if anyone raises it
